pyface/data_view/i_data_view_widget.py

```diff
@@ -8,52 +8,283 @@
+            self._set_control_selection_mode(event.new)
+
+    def _get_control_selection_mode(self):
+        """ Toolkit specific method to get the selection mode. """
+        raise NotImplementedError()
+
+    def _set_control_selection_mode(self, selection_mode):
+        """ Toolkit specific method to change the selection mode. """
+        raise NotImplementedError()
+
+    def _selection_updated(self, event):
+        """ Observer for selection trait. """
+        if self.control is not None and not self._selection_updating_flag:
+            with self._selection_updating():
+                self._set_control_selection(self.selection)
+
+    def _get_control_selection(self):
+        """ Toolkit specific method to get the selection. """
+        raise NotImplementedError()
+
+    def _set_control_selection(self, selection):
+        """ Toolkit specific method to change the selection. """
+        raise NotImplementedError()
+
+    def _observe_control_selection(self, remove=False):
+        """ Toolkit specific method to watch for changes in the selection. """
+        raise NotImplementedError()
+
+    def _update_selection(self, *args, **kwargs):
+        if not self._selection_updating_flag:
+            with self._selection_updating():
+                self._selection = self._get_control_selection()
+
+    # ------------------------------------------------------------------------
+    # Widget Interface
+    # ------------------------------------------------------------------------
+
+    def _create(self):
+        """ Creates the toolkit specific control.
+
+        This method should create the control and assign it to the
+        :py:attr:``control`` trait.
+        """
+        self.control = self._create_control(self.parent)
+        self._initialize_control()
+        self._add_event_listeners()
+
+        self.show(self.visible)
+        self.enable(self.enabled)
+
+    def _initialize_control(self):
+        """ Initializes the toolkit specific control.
+        """
+        logger.debug('Initializing DataViewWidget')
+        self._set_control_header_visible(self.header_visible)
+        self._set_control_selection_mode(self.selection_mode)
+        self._set_control_selection_type(self.selection_type)
+
     def _add_event_listeners(self):
+        logger.debug('Adding DataViewWidget listeners')
         super()._add_event_listeners()
         self.observe(
-            self._header_visible_updated, 'header_visible', dispatch='ui')
+            self._header_visible_updated,
+            'header_visible',
+            dispatch='ui',
+        )
+        self.observe(
+            self._selection_type_updated,
+            'selection_type',
+            dispatch='ui',
+        )
+        self.observe(
+            self._selection_mode_updated,
+            'selection_mode',
+            dispatch='ui',
+        )
+        self.observe(
+            self._selection_updated,
+            '_selection.items',
```

Comment:
I had a look at the trait definition approach, and it doesn't allow subclasses to easily implement additional selection modes.  As a concrete example, Qt supports a "contiguous" selection mode.  If we wanted to add that to the Qt code with this implementation, we'd just need to implement the `_set_selection` method to handle checking the validity of the new selection type (and call super() for the rest) whereas with the trait approach we would need to define a subclass of the trait type as well.

This could be fixed by delegating the validation checks back out of the trait types and onto the class, but it's getting pretty complicated.

But this is all adding complexity to try and fight traits to express what we want perfectly, but for diminishing returns.  This is a bit like the situations where you just say `foo = Any()` rather than trying to express some complex combination of traits that nails it down precisely.